File: benchmarks/pandas/bench_rolling_apply.py
"""Benchmark: rolling.apply on 10k-element pandas Series"""
import json, time
import numpy as np
import pandas as pd

ROWS = 10_000
WARMUP = 3
ITERATIONS = 10
s = pd.Series([i * 0.1 for i in range(ROWS)])

for _ in range(WARMUP):
    s.rolling(10).apply(np.mean, raw=True)

start = time.perf_counter()
for _ in range(ITERATIONS):
    s.rolling(10).apply(np.mean, raw=True)
total = (time.perf_counter() - start) * 1000
print(json.dumps({"function": "rolling_apply", "mean_ms": total / ITERATIONS, "iterations": ITERATIONS, "total_ms": total}))

File: benchmarks/tsb/bench_rolling_apply.ts
/**
 * Benchmark: rollingApply on 10k-element Series
 */
import { Series, rollingApply } from "../../src/index.js";

const ROWS = 10_000;
const WARMUP = 3;
const ITERATIONS = 10;
const data = Array.from({ length: ROWS }, (_, i) => i * 0.1);
const s = new Series({ data });
const mean = (window: number[]) => window.reduce((a, b) => a + b, 0) / window.length;

for (let i = 0; i < WARMUP; i++) rollingApply(s, 10, mean);
const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) rollingApply(s, 10, mean);
const total = performance.now() - start;
console.log(
  JSON.stringify({
    function: "rolling_apply",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
